### PR TITLE
Implement JSON Web Token support.

### PR DESCRIPTION
**Pull request by makaroshyna**

### CODE REVIEW

File: src/main/java/makaroshyna/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader(AUTHORIZATION_HEADER);
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(BEARER_TOKEN)) {
+            return bearerToken.substring(7);
```
Comment (andrii-hoienko) [inline src/main/java/makaroshyna/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
            return bearerToken.substring(BEARER_TOKEN.length);
```

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -9,3 +9,6 @@ spring.jpa.hibernate.ddl-auto=validate
 spring.jpa.show-sql=true
 spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQLDialect
 spring.jpa.open-in-view=false
+
+jwt.expiration= 300000
+jwt.secret=secretmessagemfdkfgdpwe32fosj08sjsnflwjtphsowh3ip28rnbsfk285twsdfjnweo
```
Comment (andrii-hoienko) [inline src/main/resources/application.properties:14]: Please don`t push sensitive information

File: src/test/resources/application.properties [hunk 3]
```diff
@@ -3,3 +3,6 @@ spring.datasource.driverClassName=org.h2.Driver
 spring.datasource.username=sa
 spring.datasource.password=password
 spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
+
+jwt.expiration= 10000
+jwt.secret=secretmessagemfdkfgdpwe32fosj08sjsnflwjtphsowh3ip28rnbsfk285twsdfjnweo
```
Comment (andrii-hoienko) [inline src/test/resources/application.properties:8]: Please don`t push sensitive information

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job, but some improvements required

File: src/main/java/makaroshyna/onlinebookstore/service/user/UserService.java [hunk 4]
```diff
@@ -1,8 +1,12 @@
...
+import makaroshyna.onlinebookstore.dto.user.UserLoginRequestDto;
+import makaroshyna.onlinebookstore.dto.user.UserLoginResponseDto;
 import makaroshyna.onlinebookstore.dto.user.UserRegistrationRequestDto;
 import makaroshyna.onlinebookstore.dto.user.UserRegistrationResponseDto;
 
 public interface UserService {
-    public UserRegistrationResponseDto register(UserRegistrationRequestDto requestDto);
+    UserRegistrationResponseDto register(UserRegistrationRequestDto requestDto);
+
+    UserLoginResponseDto authorize(UserLoginRequestDto requestDto);
```
Comment (d1sam) [inline src/main/java/makaroshyna/onlinebookstore/service/user/UserService.java]:
```suggestion
```

File: src/main/java/makaroshyna/onlinebookstore/dto/user/UserLoginRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+
+@Data
+public class UserLoginRequestDto {
+    @Email
+    @NotBlank
+    private String email;
```
Comment (d1sam) [inline src/main/java/makaroshyna/onlinebookstore/dto/user/UserLoginRequestDto.java:13]: what about the size?

File: src/main/java/makaroshyna/onlinebookstore/service/user/UserServiceImpl.java [hunk 6]
```diff
@@ -29,4 +37,15 @@ public UserRegistrationResponseDto register(UserRegistrationRequestDto requestDt
...
+    @Override
+    public UserLoginResponseDto authorize(UserLoginRequestDto requestDto) {
+        Authentication authentication = authenticationManager.authenticate(
+                new UsernamePasswordAuthenticationToken(
+                        requestDto.getEmail(),
+                        requestDto.getPassword())
+        );
+        String token = jwtUtil.generateToken(authentication.getName());
+        return new UserLoginResponseDto(token);
+    }
```
Comment (d1sam) [inline src/main/java/makaroshyna/onlinebookstore/service/user/UserServiceImpl.java]: let's move this method to another service `AuthenticationService` according to SRP

File: src/main/resources/application.properties (hunk 2 above)
Comment (d1sam) [reply src/main/resources/application.properties:14]: let's provide this value from env properties

File: src/test/resources/application.properties (hunk 3 above)
Comment (d1sam) [reply src/test/resources/application.properties:8]: let's provide this value from env properties

File: src/main/java/makaroshyna/onlinebookstore/controller/AuthenticationController.java [hunk 7]
```diff
@@ -24,8 +26,15 @@ public class AuthenticationController {
...
+            @RequestBody @Valid UserRegistrationRequestDto requestDto
+    ) throws RegistrationException {
         return userService.register(requestDto);
     }
+
+    @PostMapping("/login")
+    @Operation(summary = "Login a user",
+            description = "Endpoint to login users")
+    public UserLoginResponseDto login(UserLoginRequestDto requestDto) {
+        return userService.authorize(requestDto);
```
Comment (d1sam) [inline src/main/java/makaroshyna/onlinebookstore/controller/AuthenticationController.java]:
```suggestion
        return userService.authorize(requestDto);
```
please use authenticationService and call the method `authenticate`

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but there are some minor changes, that should be done according my comments.

File: src/main/java/makaroshyna/onlinebookstore/security/JwtUtil.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    public boolean isTokenValid(String token) {
+        try {
+            Jws<Claims> claimsJws = Jwts.parser()
+                    .verifyWith(secret)
+                    .build()
+                    .parseSignedClaims(token);
+            return !claimsJws.getPayload().getExpiration().before(new Date());
+        } catch (JwtException | IllegalArgumentException e) {
+            throw new JwtException("Invalid or expired JSON Web Token", e);
```
Comment (vovasalyha) [inline src/main/java/makaroshyna/onlinebookstore/security/JwtUtil.java]:
```suggestion
            return false;
```

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**